Crear validaciones y excepciones

ametrey opened the pull request:
se crean validaciones y excepciones, aún faltan alguna otras pero sirven de base

GabrielaSoruco committed:
Crear endpoint post, clase PostDTO
se creó el endpoint para agregar un nuevo Post y la clase para recibir el postDTO como argumento en el controlador y mapearlo con la clase entidad

GabrielaSoruco committed:
Agregar dependencia para la validación

ametrey committed:
Crear calidacioens y excepciones

Se crean las excepciones y validaciones basicas
para la entidad y el controller

ametrey committed:
Arreglar error de imports

Se arregla error de improts que impedia la ejecucion

ametrey committed:
Validar creacion post

Se valida la creacion del post, devuelve error 400
pero no muestra el mensaje custom

ametrey committed:
Arreglar mensaje error validaciones

en el commit anterior no devolvia el mensaje, ahora lo devuelve pero
se puede mejorar, hay que arreglar que muestra los mensajes en espanol

ametrey committed:
Arreglar mensaje resultado validacion

Se arregla el formato de mensaje ahora muestra la clave y el mesaje correspondiente

ametrey committed:
Arreglar idioma validaciones

Se crean mensajes en ingles para las validaciones

ametrey committed:
Arreglar mensajes validaciones

Se arreglan los emnsajes para mas claridad sobre los requsitos

ametrey committed:
Formatear codigo

Se da el formato correcto al codigo de varias clases

fcastellacci commented on src/main/java/mx/org/corresponsabilidadsocial/api/blog/exceptions/ErrorMessage.java:
```diff
@@ -0,0 +1,62 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class ErrorMessage {
```
No se si tiene mucha utilidad esta clase en este momento. La unica informacion que agrega es el message y para hacer eso podemos devolver un ResponseEntity<String> con el message solo. El HTTP code es parte del response

ametrey replied:
Se eliminó la clase

fcastellacci commented on src/main/java/mx/org/corresponsabilidadsocial/api/blog/exceptions/NotFoundException.java:
```diff
@@ -0,0 +1,13 @@
+public class NotFoundException extends RuntimeException {
+
+    public NotFoundException(String message) {
+        super();
```
Habria que pasarle el `message` al constructor de la clase padre

ametrey replied:
hecho, ahora el message esta en el constructor de la clase

fcastellacci commented on src/main/java/mx/org/corresponsabilidadsocial/api/blog/controllers/PostController.java:
```diff
@@ -14,16 +17,25 @@
         }
 
-        return new ResponseEntity<>(HttpStatus.NOT_FOUND);
+        throw new NotFoundException();
```
Cuando una API Rest devuelve una coleccion de entidades, esta bien devolver una lista vacia si no existen elementos. El NOT_FOUND aplica cuando estamos buscando un elemento en particular.

ametrey replied:
solucionado, devuelve una lista vacia

fcastellacci commented on src/main/java/mx/org/corresponsabilidadsocial/api/blog/controllers/PostController.java:
```diff
@@ -14,16 +17,25 @@
+    PostService postService;
 
     @GetMapping("/posts")
     public ResponseEntity<?> getPosts() throws Exception {
```
El throws aca no hace falta porque NotFoundException es una unchecked exception.
https://www.baeldung.com/java-checked-unchecked-exceptions

ametrey replied:
solucionado, se eliminó throws exception

fcastellacci commented on src/main/java/mx/org/corresponsabilidadsocial/api/blog/controllers/PostController.java:
```diff
@@ -14,16 +17,25 @@
+    @PostMapping("/newPost")
+    public ResponseEntity<?> createPost(@Valid @RequestBody Post post) {
+        if (post == null) {
+            return new ResponseEntity<>("el post esta vacío", HttpStatus.OK);
```
Habria que chequear si un post vacio pasa las validaciones que se agegaron con las annotations. De ser asi, esto deberia devolver un BAD_REQUEST ya que un post vacio es un request mal formado

ametrey replied:
cuando al post le falta el titulo, la imagen o son muy cortos devuelve un bad request con el mensaje custom

fcastellacci commented on src/main/java/mx/org/corresponsabilidadsocial/api/blog/repositories/PostRepository.java:
```diff
@@ -29,4 +29,9 @@ public void setPosts(List<Post> posts) {
     }
 
+    public void addPost(Post post) {
+        Post newPost = new Post(post.getId(), post.getTitle(), post.getImageUrl(), post.getText(), post.getStatus());
```
No es necesario crear un post nuevo, se puede pasar directamente el post a la lista

ametrey replied:
solucionado, se borra la instancia de post y directamente se agrega el psot a la lista

fcastellacci reviewed (commented):
Faltan agregar tests unitarios para las validaciones

ametrey committed:
Arreglar lo solicitado en pull request

Se borro la clase ErrorMessage y NotFoundException fue renombrada
Validaciones delegadas a PostRepository
Test de controller y service
